pkg/endpoint/policy.go

```diff
@@ -155,87 +155,143 @@ func getSecurityIdentities(owner Owner, selector *api.EndpointSelector) []policy
 				e.getLogger().WithError(err).WithField(logfields.L4PolicyID, srcID).Debug("Delete old l4 policy failed")
+				fromEndpointsSrcIDs[nipp] = false
+			} else {
+				fromEndpointsSrcIDs[nipp] = true
 			}
 		}
 	}
 
-	return errors
+	return fromEndpointsSrcIDs, errors
 }
 
-func (e *Endpoint) applyNewFilter(owner Owner, filter *policy.L4Filter) int {
+// applyNewFilter adds the given l4 filter to the endpoint and returns a map of
+// the L3-dependent L4 rules that were successfully removed if the
+// value is true, or in case of an error the value is false.
+func (e *Endpoint) applyNewFilter(owner Owner, filter *policy.L4Filter) (map[policy.NIPortProto]bool, int) {
+	fromEndpointsSrcIDs := map[policy.NIPortProto]bool{}
 	port := uint16(filter.Port)
 	proto, err := u8proto.ParseProtocol(string(filter.Protocol))
 	if err != nil {
 		e.getLogger().WithError(err).Warn("Parse policy protocol failed")
-		return 1
+		return fromEndpointsSrcIDs, 1
 	}
 
 	errors := 0
 	for _, sel := range filter.FromEndpoints {
 		for _, id := range getSecurityIdentities(owner, &sel) {
 			srcID := id.Uint32()
+			nipp := policy.NIPortProto{
+				SecID: id,
+				Port:  port,
+				Proto: uint8(proto),
+			}
 			if e.PolicyMap.L4Exists(srcID, port, uint8(proto)) {
+				fromEndpointsSrcIDs[nipp] = true
 				e.getLogger().WithField("l4Filter", filter).Debug("L4 filter exists")
 				continue
 			}
 			if err = e.PolicyMap.AllowL4(srcID, port, uint8(proto)); err != nil {
 				e.getLogger().WithError(err).Warn("Update of l4 policy map failed")
 				errors++
+				fromEndpointsSrcIDs[nipp] = false
+			} else {
+				fromEndpointsSrcIDs[nipp] = true
 			}
 		}
 	}
 
-	return errors
+	return fromEndpointsSrcIDs, errors
 }
 
 // Looks for mismatches between 'oldPolicy' and 'newPolicy', and fixes up
 // this Endpoint's BPF PolicyMap to reflect the new L3+L4 combined policy.
-func (e *Endpoint) applyL4PolicyLocked(owner Owner, oldPolicy *policy.L4Policy, newPolicy *policy.L4Policy) error {
+// Returns a map of the L3-dependent L4 rules that were successfully removed and
+// added, if the value of those maps is true, or in case of an error the value
+// is false.
+func (e *Endpoint) applyL4PolicyLocked(owner Owner, oldPolicy *policy.L4Policy,
+	newPolicy *policy.L4Policy) (map[policy.NIPortProto]bool, map[policy.NIPortProto]bool, error) {
+
 	errors := 0
+	secIDsRm := map[policy.NIPortProto]bool{}
+	secIDsAdded := map[policy.NIPortProto]bool{}
 
 	if oldPolicy != nil {
+		var secIDs map[policy.NIPortProto]bool
+		errs := 0
 		for _, filter := range oldPolicy.Ingress {
-			errors = e.removeOldFilter(owner, &filter)
+			secIDs, errs = e.removeOldFilter(owner, &filter)
+			for k, v := range secIDs {
+				e, ok := secIDsRm[k]
+				// Only set removed as true when:
+				//  - It is the first time an assignment is being done to this
+				//    key and it is true
+				//  - The previous assigned value and the new value are both
+				//    true
```

Comment:
At first read, I wasn't sure about the relationship between these comment bullets. One way that this could be improved is to use the following style:

```
Only .... when either:
- ..., OR
- ..., OR
- ...
```

Alternatively, if trying to express 'AND', the first line can stay how it is in the commit now and the "OR" can be changed to "AND".

Reply:
Fixed